terminal/upstream.go

```diff
@@ -0,0 +1,25 @@
+package terminal
+
+import "time"
+
+// Upstream defines the the interface for upstream (parent) components.
+type Upstream interface {
+	Send(msg *Msg, timeout time.Duration) *Error
+}
+
+// UpstreamFromSendFunc creates an upstream proxy from a send function.
+func UpstreamFromSendFunc(send func(msg *Msg, timeout time.Duration) *Error) *UpstreamProxy {
+	return &UpstreamProxy{
+		send: send,
+	}
+}
+
+// UpstreamProxy is a helper to be able to satisfy the Upstream interface.
+type UpstreamProxy struct {
+	send func(msg *Msg, timeout time.Duration) *Error
+}
+
+// Send is used to send a message through this upstream.
+func (up *UpstreamProxy) Send(msg *Msg, timeout time.Duration) *Error {
+	return up.send(msg, timeout)
+}
```

Comment:
This could be solved in a cleaner way: 

```golang
// Upstream defines the the interface for upstream (parent) components.
type Upstream interface {
	Send(msg *Msg, timeout time.Duration) *Error
}

type UpstreamSendFunc func(msg *Msg, timeout time.Duration) *Error

// Send is used to send a message through this upstream.
func (fn UpstreamSendFunc) Send(msg *Msg, timeout time.Duration) *Error {
	return fn(msg, timeout)
}
```

This way you just need to cast your send function with `toSomeThingWithUpstream(UpstreamSendFunc(myfn))`